optimize-onnx/src/optimize.cpp

```diff
@@ -0,0 +1,176 @@
+        }
+      }
+    }
+
+    // Additionally seed frontier with the output Values of impure
+    // Nodes.
+    for (Node * n : g->nodes()) {
+      if (isPureOperator(n)) {
+        continue;
+      }
+      live_nodes.insert(n);
+      for (Value * v : n->outputs()) {
+        live_values.insert(v);
+        frontier.push(v);
+      }
+    }
+
+    // Mark as live all Values and Nodes that are reachable from the
+    // seed Values and Nodes.
+    while (!frontier.empty()) {
+      for (Use u : frontier.front()->uses()) {
+        if (live_nodes.find(u.user) != live_nodes.end()) {
+          continue;
+        }
+        live_nodes.insert(u.user);
+        for (Value * v : u.user->outputs()) {
+          if (live_values.find(v) != live_values.end()) {
+            continue;
+          }
+          live_values.insert(v);
+          frontier.push(v);
+        }
+      }
+      frontier.pop();
+    }
+  }
+
+  if (init) {
+    // Any non-live inputs to live nodes will become an output of the
+    // init graph.
+    std::unordered_set<Value *> new_outputs;
+    for (Node * n : live_nodes) {
+      for (Value * v : n->inputs()) {
+        if (live_values.find(v) == live_values.end()) {
+          new_outputs.insert(v);
+        }
+      }
+    }
+
+    // remove all live outputs
+    for (int i = g->outputs().size() - 1; i >= 0; i--) {
+      if (live_values.find(g->outputs()[i]) != live_values.end()) {
+        g->return_node()->eraseOutput(0);
+      } else {
+        // If it's already an output that we're keeping, make sure we
+        // won't double it up.
+        new_outputs.erase(g->outputs()[i]);
+      }
+    }
+
+    // actually add in the new outputs
+    for (Value * v : new_outputs) {
+      g->registerOutput(v);
+    }
+
+    // remove all live nodes and their outputs
+    for (Node * n : live_nodes) {
+      n->destroy();
+    }
+
+    // remove all live inputs
+    for (int i = g->inputs().size() - 1; i >= 0; i--) {
+      if (live_values.find(g->inputs()[i]) != live_values.end()) {
+        g->eraseInput(i);
+      }
+    }
+  } else if (predict) {
+    // Any non-live inputs to live nodes will become an input of the
+    // predict graph.
+    std::unordered_set<Value *> new_inputs;
+    for (Node * n : live_nodes) {
+      for (Value * v : n->inputs()) {
+        if (live_values.find(v) == live_values.end()) {
+          new_inputs.insert(v);
+        }
+      }
+    }
+
+    // All other inputs can be discarded
+    for (int i = g->inputs().size() - 1; i >= 0; i--) {
+      if (live_values.find(g->inputs()[i]) != live_values.end()) {
+        // Make sure we don't double up.
+        new_inputs.erase(g->inputs()[i]);
+      } else {
+        g->eraseInput(i);
+      }
+    }
+
+    // Actually add the new inputs.
+    for (Value * v : new_inputs) {
+      g->addInput()->copyMetadata(v);
+    }
+
+    // remove all non-live nodes and their outputs.
+    for (Node * n : g->nodes()) {
+      if (n->kind() == kParam) {
```

Comment:
this shouldn't happen, right?

https://github.com/pytorch/pytorch/blob/4eb8e1276511a07ced781e7b36da39cedc4f15cc/torch/csrc/jit/ir.cpp#L432